### PR TITLE
List allowed roles per procedure in enterprise

### PR DESCRIPTION
**Pull request by Mats-SX**
The `dbms.procedures` procedure will now contain an additional
column `roles` that lists the roles that are allowed to execute it
(enterprise only). This column is only populated if the executing user
is an administrator, otherwise it will be empty.

changelog: Added column to `dbms.procedures` in enterprise, listing roles allowed to execute the procedure

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java [hunk 1]
```diff
@@ -184,19 +184,6 @@ public void resampleOutdatedIndexes()
...
-            return statement.readOperations().proceduresGetAll()
-                    .stream()
-                    .sorted( ( a, b ) -> a.name().toString().compareTo( b.name().toString() ) )
-                    .map( ProcedureResult::new );
-        }
-    }
-
     @Description( "List all user functions in the DBMS." )
     @Procedure(name = "dbms.functions", mode = READ)
     public Stream<FunctionResult> listFunctions()
```
Comment (henriknyman) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java]: Shouldn't this thinking be applied to `dbms.functions` as well?

File: enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java [hunk 2]
```diff
@@ -163,6 +168,51 @@ private KernelTransaction getCurrentTx()
...
+            {
+            case DBMS:
+                roles.add( "admin" );
+                break;
+            case READ_ONLY:
+                roles.add( "reader" );
+            case READ_WRITE:
+                roles.add( "publisher" );
+            case SCHEMA_WRITE:
+                roles.add( "architect" );
```
Comment (fickludd) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java]: So will this switch add the architect role to a `mode=READ` query? Do we want that?

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInDbmsProcedures.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import org.neo4j.kernel.impl.proc.Procedures;
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.procedure.Context;
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Procedure;
+
+import static org.neo4j.procedure.Mode.DBMS;
+
+@SuppressWarnings( "unused" )
+public class BuiltInDbmsProcedures
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInDbmsProcedures.java:35]: I think I see what you are doing here, but then maybe we could have the same class-name in enterprise for consistency? Atm. enterprise listProcedures is in `BuiltInProcedures`, while community lies in `BuiltInDbmsProcedures`.

File: enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ConfiguredProceduresTestBase.java [hunk 4]
```diff
@@ -222,4 +233,48 @@ public void shouldHandleMultipleRolesSpecifiedForMapping() throws Throwable
...
+                "db.labels", newSet( "default", READER, PUBLISHER, ARCHITECT, ADMIN ),
+                "dbms.security.changePassword", newSet( ADMIN ),
+                "dbms.procedures", newSet( ADMIN ),
+                "dbms.listQueries", newSet( ADMIN ),
+                "dbms.security.createUser", newSet( ADMIN ) );
+
+        assertListProceduresHasRoles( adminSubject, expected );
+        assertListProceduresHasRoles( schemaSubject, expected );
+        assertListProceduresHasRoles( writeSubject, expected );
+        assertListProceduresHasRoles( writeSubject, expected );
```
Comment (fickludd) [inline enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ConfiguredProceduresTestBase.java]: Over-assertion!

**Comment by fickludd** [comment]
I think this is good.

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java (hunk 1 above)
Comment (Mats-SX) [reply community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java]: You're right, that would make sense.

**Pull request merged**